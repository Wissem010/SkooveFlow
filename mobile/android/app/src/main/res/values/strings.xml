<resources>
    <string name="app_name">SkooveFlow</string>
</resources>
