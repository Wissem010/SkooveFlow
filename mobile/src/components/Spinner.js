import {View, ActivityIndicator, StyleSheet} from 'react-native';
import React from 'react';

export default function Spinner() {
  return (
    <View style={styles.spinnerContainer}>
      <ActivityIndicator size={'large'} color={'grey'} />
    </View>
  );
}

const styles = StyleSheet.create({
  spinnerContainer: {
    height: '100%',
    display: 'flex',
    alignContent: 'center',
    justifyContent: 'center',
  },
});
