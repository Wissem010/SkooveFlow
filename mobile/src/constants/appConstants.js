export const radioButtonsData = [
  {id: 1, label: 'option 1', value: 'option1'},
  {id: 2, label: 'option 2', value: 'option2'},
  {id: 3, label: 'option 3', value: 'option3'},
  {id: 4, label: 'option 4', value: 'option4'},
  {id: 5, label: 'option 5', value: 'option5'},
];
