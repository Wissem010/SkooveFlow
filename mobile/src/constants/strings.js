const SCREEN_A = 'ScreenA';
const SCREEN_B = 'ScreenB';
const SCREEN_D = 'ScreenD';
const RETRY_MESSAGE = 'Try resubmitting your answer...';
const REDIRECT_MESSAGE = "You'll be redirected shortly...";
const RADIO_GROUP_SELECTION_ERR_MSG = 'Please selection an item from the list';

module.exports = {
  SCREEN_A: SCREEN_A,
  SCREEN_B: SCREEN_B,
  SCREEN_D: SCREEN_D,
  RETRY_MESSAGE: RETRY_MESSAGE,
  REDIRECT_MESSAGE: REDIRECT_MESSAGE,
  RADIO_GROUP_SELECTION_ERR_MSG: RADIO_GROUP_SELECTION_ERR_MSG,
};
