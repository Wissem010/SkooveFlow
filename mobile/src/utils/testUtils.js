import React from 'react';
import {render} from '@testing-library/react';
import {configureStore} from '@reduxjs/toolkit';
import {Provider} from 'react-redux';
import {expirementReducer} from '../redux/slices/expirementSlice';
import {selectionReducer} from '../redux/slices/selectionSlice';

export function renderWithProviders(
  ui,
  {
    preloadedState = {},
    store = configureStore({
      reducer: {
        expirementReducer: expirementReducer,
        selectionReducer: selectionReducer,
      },
      preloadedState,
    }),
    ...renderOptions
  } = {},
) {
  function Wrapper({children}) {
    return <Provider store={store}>{children}</Provider>;
  }

  return {store, ...render(ui, {wrapper: Wrapper, ...renderOptions})};
}
