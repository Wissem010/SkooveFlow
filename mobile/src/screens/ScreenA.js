//React imports
import React, {useEffect} from 'react';
//Redux imports
import {useDispatch, useSelector} from 'react-redux';
import {
  resetScreenAState,
  rFetchExpirements,
} from '../redux/slices/expirementSlice';
//Ui imports
import Spinner from '../components/Spinner';
import {View, StyleSheet, Text} from 'react-native';
//Constants imports
import {SCREEN_B, SCREEN_D} from '../constants/strings';

export default function ScreenA({navigation}) {
  const dispatch = useDispatch();
  const {loading, nextScreen, err, message} = useSelector(
    state => state.expirementReducer,
  );
  useEffect(() => {
    return navigation.addListener('focus', () => {
      dispatch(rFetchExpirements());
    });
  }, [navigation]);

  useEffect(() => {
    if (nextScreen == SCREEN_B) {
      navigation.navigate(SCREEN_B);
      dispatch(resetScreenAState());
    }
    if (nextScreen == SCREEN_D) {
      setTimeout(() => {
        navigation.navigate(SCREEN_D);
        dispatch(resetScreenAState());
      }, 5000);
    }
  }, [nextScreen]);

  return (
    <View style={styles.backgroundStyle}>
      {loading ? (
        <Spinner />
      ) : (
        err && <Text style={styles.messageStyle}>{message}</Text>
      )}
    </View>
  );
}

const styles = StyleSheet.create({
  backgroundStyle: {
    backgroundColor: 'lightgreen',
    height: '100%',
    display: 'flex',
    alignContent: 'center',
    justifyContent: 'center',
  },
  messageStyle: {
    alignSelf: 'center',
    fontSize: 28,
  },
});
