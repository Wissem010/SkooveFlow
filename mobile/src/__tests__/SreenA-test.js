import React from 'react';
import renderer from 'react-test-renderer';
import Spinner from '../components/Spinner';
import * as reactRedux from 'react-redux';

import {REDIRECT_MESSAGE, SCREEN_B, SCREEN_D} from '../constants/strings';
import {
  expirementReducer,
  resetScreenAState,
  rFetchExpirements,
} from '../redux/slices/expirementSlice';
import ScreenA from '../screens/ScreenA';
import {renderWithProviders} from '../utils/testUtils';

//Mocking navigation functions
const mockNavigation = {
  navigate: jest.fn(),
  addListener: jest.fn(),
};

//Mocking redux functions
jest.mock('react-redux', () => ({
  useSelector: jest.fn(),
  useDispatch: jest.fn(),
}));

const mockStore = {
  expirementReducer,
};
const useSelectorMock = reactRedux.useSelector;
const useDispatchMock = reactRedux.useDispatch;

//Initializing mocks
beforeEach(() => {
  useDispatchMock.mockImplementation(() => () => {});
  useSelectorMock.mockImplementation(selector => selector(mockStore));
});
afterEach(() => {
  useDispatchMock.mockClear();
  useSelectorMock.mockClear();
});

//Tests
test('renders correctly', () => {
  const tree = renderer.create(<Spinner />).toJSON();
  expect(tree).toMatchSnapshot();
});

test('should return the fetchExpirement initial state', () => {
  expect(expirementReducer(undefined, {type: undefined})).toEqual({
    loading: false,
    message: '',
    err: false,
    nextScreen: '',
  });
});

test('should return the wanted state after dispatch resetScreenA action', () => {
  expect(expirementReducer(undefined, resetScreenAState)).toEqual({
    loading: false,
    message: '',
    err: false,
    nextScreen: '',
  });
});

describe('SELECTION SLICE TESTS', () => {
  it('should set loading true while action is pending', () => {
    const action = {type: rFetchExpirements.pending};
    const state = expirementReducer(
      {
        loading: false,
        message: '',
        err: false,
        nextScreen: '',
      },
      action,
    );
    expect(state).toEqual({
      loading: true,
      message: '',
      err: false,
      nextScreen: '',
    });
  });

  it('should set correct screen when action is fulfilled', () => {
    const action = {
      type: rFetchExpirements.fulfilled,
    };
    const state = expirementReducer(
      {
        loading: false,
        message: '',
        err: false,
        nextScreen: '',
      },
      action,
    );
    expect(state).toEqual({
      loading: false,
      message: '',
      err: false,
      nextScreen: SCREEN_B,
    });
  });

  it('should set error true when action is rejected', () => {
    const action = {type: rFetchExpirements.rejected};
    const state = expirementReducer(
      {
        loading: false,
        message: '',
        err: false,
        nextScreen: '',
      },
      action,
    );
    expect(state).toEqual({
      loading: false,
      message: REDIRECT_MESSAGE,
      err: true,
      nextScreen: SCREEN_D,
    });
  });
});

test('renders screenA with providers correctly', () => {
  const tree = renderWithProviders(
    <ScreenA navigation={mockNavigation} />,
  ).toJSON();
  expect(tree).toMatchSnapshot();
});
