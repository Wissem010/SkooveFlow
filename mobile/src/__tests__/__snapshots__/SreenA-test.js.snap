// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<View
  style={
    Object {
      "alignContent": "center",
      "display": "flex",
      "height": "100%",
      "justifyContent": "center",
    }
  }
>
  <ActivityIndicator
    color="grey"
    size="large"
  />
</View>
`;

exports[`renders screenA with providers correctly 1`] = `
<View
  style={
    Object {
      "alignContent": "center",
      "backgroundColor": "lightgreen",
      "display": "flex",
      "height": "100%",
      "justifyContent": "center",
    }
  }
/>
`;
